lib/autoupdate/agent/process.go

```diff
@@ -109,27 +117,67 @@ func (s SystemdService) Reload(ctx context.Context) error {
 	default:
 		s.Log.InfoContext(ctx, "Gracefully reloaded.", unitKey, s.ServiceName)
 	}
+	// monitor logs all relevant errors, so we filter for a few outcomes
+	err = s.monitor(ctx, initPID)
+	if errors.Is(err, context.DeadlineExceeded) {
+		return trace.Errorf("timed out while waiting for process to start")
+	}
+	if errors.Is(err, context.Canceled) {
+		return context.Canceled
+	}
+	if err != nil {
+		return trace.Errorf("failed to monitor process")
+	}
+	return nil
+}
+
+// monitor for a started, healthy process.
+// monitor logs all errors that should be displayed to the user.
+func (s SystemdService) monitor(ctx context.Context, initPID int) error {
+	ctx, cancel := context.WithTimeout(ctx, monitorTimeout)
+	defer cancel()
+	tickC := time.NewTicker(monitorInterval).C
+
 	if initPID != 0 {
 		s.Log.InfoContext(ctx, "Monitoring PID file to detect crashes.", unitKey, s.ServiceName)
-		err := s.monitor(ctx, initPID)
-		if errors.Is(err, context.DeadlineExceeded) {
-			return trace.Errorf("timed out while waiting for process to start")
+		err := s.monitorPID(ctx, initPID, tickC)
+		if err != nil {
+			s.Log.ErrorContext(ctx, "Error monitoring for crashing PID.", errorKey, err, unitKey, s.ServiceName)
+			return trace.Wrap(err)
 		}
+	}
+
+	s.Log.InfoContext(ctx, "Monitoring diagnostic socket to detect readiness.", unitKey, s.ServiceName)
+	client := debug.NewClient(s.SocketPath)
+	err := s.waitForSocketPresent(ctx, tickC, missingSocketTimeout)
```

Comment:
What's the point of statting the socket? Can't we just connect to it? I'm pretty sure that DialUnix will return os.ErrNotExist on a missing socket and syscall.EINVAL on a non-socket, so we can still distinguish the two cases we explicitly single out in `waitForSocketPresent`.

Reply:
The point is just to wait for the socket to be created by the new process. Happy to combine this with the polling in `waitForReady` though.